finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,675 @@
+        test_network,
+    );
+
+    let chain_builder = ClientChainBuilder::new(client, Arc::new(TestClientBuilder::new().build()));
+    let (exit_data_store_tx, exit_data_store_rx) = oneshot::channel();
+
+    (
+        async move {
+            data_store.run(exit_data_store_rx).await;
+        },
+        exit_data_store_tx,
+        TestHandler {
+            chain_builder,
+            block_requests_rx,
+            network_tx,
+            network: Box::new(network),
+        },
+    )
+}
+
+const TIMEOUT_SUCC: Duration = Duration::from_millis(5000);
+const TIMEOUT_FAIL: Duration = Duration::from_millis(200);
+
+// This is the basic assumption for other tests, so we better test it, in case this somehow changes in the future.
+#[tokio::test]
+async fn forks_have_different_block_hashes() {
+    let (_task_handle, _exit, mut test_handler) = prepare_data_store(None);
+    let genesis_hash = test_handler.genesis_hash();
+    let a1 = test_handler.build_block_above(&genesis_hash).await;
+    let b1 = test_handler.build_block_above(&genesis_hash).await;
+    assert_ne!(a1.hash(), b1.hash());
+}
+
+async fn run_test<F, S>(scenario: S)
+where
+    F: Future,
+    S: FnOnce(TestHandler) -> F,
+{
+    let (task_handle, exit, test_handler) = prepare_data_store(None);
+    let data_store_handle = tokio::spawn(task_handle);
+
+    scenario(test_handler).await;
+
+    exit.send(()).unwrap();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn correct_messages_go_through() {
+    run_test(|mut test_handler| async move {
+        let blocks = test_handler
+            .initialize_single_branch_and_import(MAX_DATA_BRANCH_LEN * 10)
+            .await;
+
+        for i in 1..=MAX_DATA_BRANCH_LEN {
+            let blocks_branch = blocks[0..(i as usize)].to_vec();
+            let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+            test_handler.send_data(test_data.clone());
+
+            let message = test_handler
+                .assert_message_out("Did not receive message from Data Store")
+                .await;
+            assert_eq!(message.included_data(), test_data);
+        }
+    })
+    .await;
+}
+
+#[tokio::test]
+async fn too_long_branch_message_does_not_go_through() {
+    run_test(|mut test_handler| async move {
+        let blocks = test_handler
+            .initialize_single_branch_and_import(MAX_DATA_BRANCH_LEN * 10)
+            .await;
+
+        test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 2].hash());
+
+        let blocks_branch = blocks[0..((MAX_DATA_BRANCH_LEN + 1) as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+        test_handler
+            .assert_no_message_out("Data Store let through a too long message")
+            .await;
+    })
+    .await;
+}
+
+#[tokio::test]
+async fn branch_not_within_session_boundaries_does_not_go_through() {
```

Comment:
why don't we use `run_test` here?

Reply:
This has non-standard session boundaries. Shall I add session_boundaries as an argument to `run_test` -- then I can use `run_test` here.

Reply:
maybe just let's ignore this